File: vss-android/app/src/main/java/com/vss/MediaFrameSource.java
package com.vss;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.ImageDecoder;
import android.graphics.ImageFormat;
import android.hardware.DataSpace;
import android.hardware.HardwareBuffer;
import android.hardware.SyncFence;
import android.media.Image;
import android.media.ImageReader;
import android.media.MediaCodec;
import android.media.MediaExtractor;
import android.media.MediaFormat;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.util.Log;

import com.vss.simulator.SimulatorSurfaceView;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.concurrent.CountDownLatch;

class MediaFrameSource {
    private static final String LOG_TAG = "MediaFrameSource";
    private static final int MAX_IMAGE_EDGE = 1920;
    private static final int MAX_VIDEO_IMAGES = 6;
    private static final int SURFACE_RETRY_DELAY_MS = 16;
    private static final long CODEC_TIMEOUT_US = 10_000L;

    private final Context context;
    private final SimulatorSurfaceView simulatorView;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final Object hardwareFrameLock = new Object();

    private Thread workerThread;
    private volatile boolean cancelled;
    private HardwareFrame pendingHardwareFrame;
    private boolean hardwareFramePostScheduled;

    MediaFrameSource(Context context, SimulatorSurfaceView simulatorView) {
        this.context = context.getApplicationContext();
        this.simulatorView = simulatorView;
    }

    void start(Uri uri, String mimeType) {
        close();
        cancelled = false;
        workerThread = new Thread(() -> decode(uri, mimeType), "VSS media");
        workerThread.start();
    }

    void close() {
        cancelled = true;
        closePendingHardwareFrame();
        if (workerThread != null) {
            workerThread.interrupt();
            workerThread = null;
        }
    }

    private void decode(Uri uri, String mimeType) {
        try {
            String resolvedMimeType = resolveMimeType(uri, mimeType);
            if (resolvedMimeType != null && resolvedMimeType.startsWith("video/")) {
                decodeVideo(uri);
            } else if (!decodeImage(uri)) {
                Log.w(LOG_TAG, "Image could not be decoded: " + uri);
            }
        } catch (Exception err) {
            Log.e(LOG_TAG, "Cannot decode media " + uri, err);
        }
    }

    private String resolveMimeType(Uri uri, String mimeType) {
        if (mimeType != null && !mimeType.equals("*/*")) {
            return mimeType;
        }
        return context.getContentResolver().getType(uri);
    }

    private boolean decodeImage(Uri uri) throws IOException {
        Bitmap bitmap;
        try {
            ImageDecoder.Source source = ImageDecoder.createSource(context.getContentResolver(), uri);
            bitmap = ImageDecoder.decodeBitmap(source, (decoder, info, src) -> {
                decoder.setAllocator(ImageDecoder.ALLOCATOR_SOFTWARE);
                int width = info.getSize().getWidth();
                int height = info.getSize().getHeight();
                int sampleSize = sampleSize(width, height);
                if (sampleSize > 1) {
                    decoder.setTargetSize(Math.max(1, width / sampleSize), Math.max(1, height / sampleSize));
                }
            });
        } catch (IOException | RuntimeException err) {
            Log.w(LOG_TAG, "Hardware image decode failed", err);
            return false;
        } catch (OutOfMemoryError err) {
            Log.w(LOG_TAG, "Hardware image decode ran out of memory", err);
            return false;
        }

        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        ByteBuffer pixels = ByteBuffer.allocateDirect(width * height * 4);
        bitmap.copyPixelsToBuffer(pixels);
        pixels.rewind();
        bitmap.recycle();
        mainHandler.post(() -> postRgbaWhenSurfaceIsReady(width, height, pixels));
        return true;
    }

    private void postRgbaWhenSurfaceIsReady(int width, int height, ByteBuffer pixels) {
        if (cancelled) {
            return;
        }
        if (!simulatorView.getHolder().getSurface().isValid()) {
            mainHandler.postDelayed(
                    () -> postRgbaWhenSurfaceIsReady(width, height, pixels),
                    SURFACE_RETRY_DELAY_MS
            );
            return;
        }
        simulatorView.postRgba(width, height, pixels);
    }

    private int sampleSize(int width, int height) {
        int sampleSize = 1;
        int edge = Math.max(width, height);
        while (edge / sampleSize > MAX_IMAGE_EDGE) {
            sampleSize *= 2;
        }
        return sampleSize;
    }

    private void decodeVideo(Uri uri) throws IOException {
        MediaExtractor extractor = new MediaExtractor();
        MediaCodec decoder = null;
        ImageReader imageReader = null;
        HandlerThread imageThread = null;
        boolean decoderStarted = false;
        try {
            extractor.setDataSource(context, uri, null);
            int trackIndex = selectVideoTrack(extractor);
            if (trackIndex < 0) {
                Log.w(LOG_TAG, "Selected media contains no video track: " + uri);
                return;
            }

            extractor.selectTrack(trackIndex);
            MediaFormat format = extractor.getTrackFormat(trackIndex);
            String mimeType = format.getString(MediaFormat.KEY_MIME);
            if (mimeType == null) {
                Log.w(LOG_TAG, "Selected video track has no MIME type: " + uri);
                return;
            }

            int width = format.getInteger(MediaFormat.KEY_WIDTH);
            int height = format.getInteger(MediaFormat.KEY_HEIGHT);
            int rotationDegrees = videoRotationDegrees(format);

            imageThread = new HandlerThread("VSS media images");
            imageThread.start();
            Handler imageHandler = new Handler(imageThread.getLooper());
            imageReader = ImageReader.newInstance(width, height, ImageFormat.PRIVATE, MAX_VIDEO_IMAGES);
            final ImageReader frameReader = imageReader;
            final int frameWidth = width;
            final int frameHeight = height;
            final int frameRotationDegrees = rotationDegrees;
            imageReader.setOnImageAvailableListener(reader -> {
                Image image;
                try {
                    image = reader.acquireLatestImage();
                } catch (IllegalStateException err) {
                    Log.w(LOG_TAG, "Dropping decoded frame because the media ImageReader is full", err);
                    return;
                }
                if (image == null) {
                    return;
                }

                HardwareBuffer hardwareBuffer = null;
                try {
                    waitForAcquireFence(image);
                    hardwareBuffer = image.getHardwareBuffer();
                    if (hardwareBuffer == null) {
                        image.close();
                        return;
                    }
                    submitHardwareFrame(
                            frameWidth,
                            frameHeight,
                            image.getDataSpace(),
                            frameRotationDegrees,
                            hardwareBuffer,
                            image::close
                    );
                } catch (RuntimeException err) {
                    Log.w(LOG_TAG, "Cannot forward decoded video frame", err);
                    if (hardwareBuffer != null) {
                        hardwareBuffer.close();
                    }
                    image.close();
                }
            }, imageHandler);

            decoder = MediaCodec.createDecoderByType(mimeType);
            decoder.configure(format, frameReader.getSurface(), null, 0);
            decoder.start();
            decoderStarted = true;
            drainVideo(extractor, decoder, imageHandler);
        } finally {
            if (decoder != null) {
                if (decoderStarted) {
                    decoder.stop();
                }
                decoder.release();
            }
            if (imageReader != null) {
                imageReader.close();
            }
            if (imageThread != null) {
                imageThread.quitSafely();
            }
            extractor.release();
        }
    }

    private int selectVideoTrack(MediaExtractor extractor) {
        for (int i = 0; i < extractor.getTrackCount(); i++) {
            MediaFormat format = extractor.getTrackFormat(i);
            String mimeType = format.getString(MediaFormat.KEY_MIME);
            if (mimeType != null && mimeType.startsWith("video/")) {
                return i;
            }
        }
        return -1;
    }

    private int videoRotationDegrees(MediaFormat format) {
        if (!format.containsKey(MediaFormat.KEY_ROTATION)) {
            return 0;
        }
        return normalizeRotationDegrees(format.getInteger(MediaFormat.KEY_ROTATION));
    }

    private int normalizeRotationDegrees(int rotationDegrees) {
        int normalized = ((rotationDegrees % 360) + 360) % 360;
        if (normalized >= 45 && normalized <= 134) {
            return 90;
        }
        if (normalized >= 135 && normalized <= 224) {
            return 180;
        }
        if (normalized >= 225 && normalized <= 314) {
            return 270;
        }
        return 0;
    }

    private void drainVideo(MediaExtractor extractor, MediaCodec decoder, Handler imageHandler) {
        MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();
        boolean inputEnded = false;
        long playbackStartMs = -1L;

        while (!cancelled) {
            if (!inputEnded) {
                int inputIndex = decoder.dequeueInputBuffer(CODEC_TIMEOUT_US);
                if (inputIndex >= 0) {
                    ByteBuffer inputBuffer = decoder.getInputBuffer(inputIndex);
                    if (inputBuffer != null) {
                        inputBuffer.clear();
                    }
                    int sampleSize = inputBuffer != null
                            ? extractor.readSampleData(inputBuffer, 0)
                            : -1;
                    if (sampleSize < 0) {
                        decoder.queueInputBuffer(
                                inputIndex,
                                0,
                                0,
                                0,
                                MediaCodec.BUFFER_FLAG_END_OF_STREAM
                        );
                        inputEnded = true;
                    } else {
                        decoder.queueInputBuffer(
                                inputIndex,
                                0,
                                sampleSize,
                                extractor.getSampleTime(),
                                extractor.getSampleFlags()
                        );
                        extractor.advance();
                    }
                }
            }

            int outputIndex = decoder.dequeueOutputBuffer(bufferInfo, CODEC_TIMEOUT_US);
            if (outputIndex >= 0) {
                if (playbackStartMs < 0) {
                    playbackStartMs = System.currentTimeMillis() - bufferInfo.presentationTimeUs / 1000L;
                }
                waitUntilPresentationTime(playbackStartMs, bufferInfo.presentationTimeUs);
                boolean render = (bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) == 0;
                decoder.releaseOutputBuffer(outputIndex, render);
                if ((bufferInfo.flags & MediaCodec.BUFFER_FLAG_END_OF_STREAM) != 0) {
                    waitForPendingVideoFrames(imageHandler);
                    decoder.flush();
                    extractor.seekTo(0, MediaExtractor.SEEK_TO_PREVIOUS_SYNC);
                    inputEnded = false;
                    playbackStartMs = -1L;
                }
            } else if (outputIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                Log.d(LOG_TAG, "Video decoder output format changed: " + decoder.getOutputFormat());
            }
        }
    }

    private void waitForPendingVideoFrames(Handler imageHandler) {
        CountDownLatch imagesDrained = new CountDownLatch(1);
        CountDownLatch framesSubmitted = new CountDownLatch(1);
        imageHandler.post(() -> {
            imagesDrained.countDown();
            mainHandler.post(framesSubmitted::countDown);
        });
        try {
            imagesDrained.await();
            framesSubmitted.await();
        } catch (InterruptedException err) {
            Thread.currentThread().interrupt();
            cancelled = true;
        }
    }

    private void waitUntilPresentationTime(long playbackStartMs, long presentationTimeUs) {
        long delayMs = playbackStartMs + presentationTimeUs / 1000L - System.currentTimeMillis();
        if (delayMs <= 0) {
            return;
        }
        try {
            Thread.sleep(delayMs);
        } catch (InterruptedException err) {
            Thread.currentThread().interrupt();
            cancelled = true;
        }
    }

    private void submitHardwareFrame(
            int width,
            int height,
            int dataSpace,
            int rotationDegrees,
            HardwareBuffer hardwareBuffer,
            Runnable releaseOwner
    ) {
        postHardwareFrameWhenSurfaceIsReady(new HardwareFrame(
                width,
                height,
                dataSpace,
                rotationDegrees,
                hardwareBuffer,
                releaseOwner
        ));
    }

    private void postHardwareFrameWhenSurfaceIsReady(HardwareFrame frame) {
        boolean shouldPost = false;
        synchronized (hardwareFrameLock) {
            if (pendingHardwareFrame != null) {
                pendingHardwareFrame.close();
            }
            pendingHardwareFrame = frame;
            if (!hardwareFramePostScheduled) {
                hardwareFramePostScheduled = true;
                shouldPost = true;
            }
        }
        if (shouldPost) {
            mainHandler.post(this::postPendingHardwareFrameOnMainThread);
        }
    }

    private void postPendingHardwareFrameOnMainThread() {
        HardwareFrame frame;
        synchronized (hardwareFrameLock) {
            frame = pendingHardwareFrame;
            pendingHardwareFrame = null;
            if (frame == null) {
                hardwareFramePostScheduled = false;
                return;
            }
        }

        if (cancelled) {
            frame.close();
            synchronized (hardwareFrameLock) {
                hardwareFramePostScheduled = false;
            }
            return;
        }
        if (!simulatorView.getHolder().getSurface().isValid()) {
            mainHandler.postDelayed(
                    this::postPendingHardwareFrameOnMainThread,
                    SURFACE_RETRY_DELAY_MS
            );
            synchronized (hardwareFrameLock) {
                if (pendingHardwareFrame != null) {
                    frame.close();
                } else {
                    pendingHardwareFrame = frame;
                }
            }
            return;
        }
        try {
            simulatorView.postHardwareBuffer(
                    frame.width,
                    frame.height,
                    frame.dataSpace,
                    frame.rotationDegrees,
                    frame.hardwareBuffer
            );
        } finally {
            frame.close();
        }

        boolean shouldPostAgain;
        synchronized (hardwareFrameLock) {
            shouldPostAgain = pendingHardwareFrame != null;
            hardwareFramePostScheduled = shouldPostAgain;
        }
        if (shouldPostAgain) {
            mainHandler.post(this::postPendingHardwareFrameOnMainThread);
        }
    }

    private void closePendingHardwareFrame() {
        synchronized (hardwareFrameLock) {
            if (pendingHardwareFrame != null) {
                pendingHardwareFrame.close();
                pendingHardwareFrame = null;
            }
            hardwareFramePostScheduled = false;
        }
    }

    private void waitForAcquireFence(Image image) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) {
            return;
        }

        try (SyncFence fence = image.getFence()) {
            if (fence != null && fence.isValid() && !fence.awaitForever()) {
                Log.w(LOG_TAG, "Media acquire fence reported an error");
            }
        } catch (IOException e) {
            Log.w(LOG_TAG, "Cannot get media acquire fence; continuing without explicit wait", e);
        }
    }

    private static class HardwareFrame {
        final int width;
        final int height;
        final int dataSpace;
        final int rotationDegrees;
        final HardwareBuffer hardwareBuffer;
        final Runnable releaseOwner;

        HardwareFrame(
                int width,
                int height,
                int dataSpace,
                int rotationDegrees,
                HardwareBuffer hardwareBuffer,
                Runnable releaseOwner
        ) {
            this.width = width;
            this.height = height;
            this.dataSpace = dataSpace;
            this.rotationDegrees = rotationDegrees;
            this.hardwareBuffer = hardwareBuffer;
            this.releaseOwner = releaseOwner;
        }

        void close() {
            try {
                hardwareBuffer.close();
            } finally {
                releaseOwner.run();
            }
        }
    }

}

File: vss-android/app/src/main/kotlin/com/vss/MediaFrameSource.kt
package com.vss

import android.content.Context
import android.graphics.Bitmap
import android.graphics.ImageDecoder
import android.graphics.ImageFormat
import android.hardware.HardwareBuffer
import android.hardware.SyncFence
import android.media.Image
import android.media.ImageReader
import android.media.MediaCodec
import android.media.MediaExtractor
import android.media.MediaFormat
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import java.io.IOException
import java.nio.ByteBuffer
import java.util.concurrent.CountDownLatch
import kotlin.math.max

class MediaFrameSource(context: Context, private val controller: SimulatorController) {
    private val context = context.applicationContext
    private val mainHandler = Handler(Looper.getMainLooper())
    private val hardwareFrameLock = Any()

    private var workerThread: Thread? = null

    @Volatile
    private var cancelled = false

    private var pendingHardwareFrame: HardwareFrame? = null
    private var hardwareFramePostScheduled = false

    fun start(uri: Uri, mimeType: String?) {
        close()
        cancelled = false
        workerThread = Thread({ decode(uri, mimeType) }, "VSS media").also { it.start() }
    }

    fun close() {
        cancelled = true
        closePendingHardwareFrame()
        workerThread?.interrupt()
        workerThread = null
    }

    private fun decode(uri: Uri, mimeType: String?) {
        try {
            val resolvedMimeType = resolveMimeType(uri, mimeType)
            if (resolvedMimeType?.startsWith("video/") == true) {
                decodeVideo(uri)
            } else if (!decodeImage(uri)) {
                Log.w(LOG_TAG, "Image could not be decoded: $uri")
            }
        } catch (error: Exception) {
            Log.e(LOG_TAG, "Cannot decode media $uri", error)
        }
    }

    private fun resolveMimeType(uri: Uri, mimeType: String?): String? {
        if (mimeType != null && mimeType != "*/*") return mimeType
        return context.contentResolver.getType(uri)
    }

    @Throws(IOException::class)
    private fun decodeImage(uri: Uri): Boolean {
        val bitmap: Bitmap
        try {
            val source = ImageDecoder.createSource(context.contentResolver, uri)
            bitmap = ImageDecoder.decodeBitmap(source) { decoder, info, _ ->
                decoder.allocator = ImageDecoder.ALLOCATOR_SOFTWARE
                val width = info.size.width
                val height = info.size.height
                val sampleSize = sampleSize(width, height)
                if (sampleSize > 1) {
                    decoder.setTargetSize(
                        max(1, width / sampleSize),
                        max(1, height / sampleSize),
                    )
                }
            }
        } catch (error: IOException) {
            Log.w(LOG_TAG, "Hardware image decode failed", error)
            return false
        } catch (error: RuntimeException) {
            Log.w(LOG_TAG, "Hardware image decode failed", error)
            return false
        } catch (error: OutOfMemoryError) {
            Log.w(LOG_TAG, "Hardware image decode ran out of memory", error)
            return false
        }

        val width = bitmap.width
        val height = bitmap.height
        val pixels = ByteBuffer.allocateDirect(width * height * 4)
        bitmap.copyPixelsToBuffer(pixels)
        pixels.rewind()
        bitmap.recycle()
        mainHandler.post { postRgbaWhenSurfaceIsReady(width, height, pixels) }
        return true
    }

    private fun postRgbaWhenSurfaceIsReady(width: Int, height: Int, pixels: ByteBuffer) {
        if (cancelled) return
        if (!controller.isReady()) {
            mainHandler.postDelayed(
                { postRgbaWhenSurfaceIsReady(width, height, pixels) },
                SURFACE_RETRY_DELAY_MS,
            )
            return
        }
        controller.postRgba(width, height, pixels)
    }

    private fun sampleSize(width: Int, height: Int): Int {
        var sampleSize = 1
        val edge = max(width, height)
        while (edge / sampleSize > MAX_IMAGE_EDGE) {
            sampleSize *= 2
        }
        return sampleSize
    }

    @Throws(IOException::class)
    private fun decodeVideo(uri: Uri) {
        val extractor = MediaExtractor()
        var decoder: MediaCodec? = null
        var imageReader: ImageReader? = null
        var imageThread: HandlerThread? = null
        var decoderStarted = false
        try {
            extractor.setDataSource(context, uri, null)
            val trackIndex = selectVideoTrack(extractor)
            if (trackIndex < 0) {
                Log.w(LOG_TAG, "Selected media contains no video track: $uri")
                return
            }

            extractor.selectTrack(trackIndex)
            val format = extractor.getTrackFormat(trackIndex)
            val mimeType = format.getString(MediaFormat.KEY_MIME)
            if (mimeType == null) {
                Log.w(LOG_TAG, "Selected video track has no MIME type: $uri")
                return
            }

            val width = format.getInteger(MediaFormat.KEY_WIDTH)
            val height = format.getInteger(MediaFormat.KEY_HEIGHT)
            val rotationDegrees = videoRotationDegrees(format)

            imageThread = HandlerThread("VSS media images").also { it.start() }
            val imageHandler = Handler(imageThread.looper)
            imageReader = ImageReader.newInstance(
                width,
                height,
                ImageFormat.PRIVATE,
                MAX_VIDEO_IMAGES,
            )
            val frameReader = imageReader
            imageReader.setOnImageAvailableListener({ reader ->
                val image = try {
                    reader.acquireLatestImage()
                } catch (error: IllegalStateException) {
                    Log.w(
                        LOG_TAG,
                        "Dropping decoded frame because the media ImageReader is full",
                        error,
                    )
                    return@setOnImageAvailableListener
                } ?: return@setOnImageAvailableListener

                var hardwareBuffer: HardwareBuffer? = null
                try {
                    waitForAcquireFence(image)
                    hardwareBuffer = image.hardwareBuffer
                    if (hardwareBuffer == null) {
                        image.close()
                        return@setOnImageAvailableListener
                    }
                    submitHardwareFrame(
                        width,
                        height,
                        image.dataSpace,
                        rotationDegrees,
                        hardwareBuffer,
                        image::close,
                    )
                } catch (error: RuntimeException) {
                    Log.w(LOG_TAG, "Cannot forward decoded video frame", error)
                    hardwareBuffer?.close()
                    image.close()
                }
            }, imageHandler)

            decoder = MediaCodec.createDecoderByType(mimeType)
            decoder.configure(format, frameReader.surface, null, 0)
            decoder.start()
            decoderStarted = true
            drainVideo(extractor, decoder, imageHandler)
        } finally {
            decoder?.let {
                if (decoderStarted) it.stop()
                it.release()
            }
            imageReader?.close()
            imageThread?.quitSafely()
            extractor.release()
        }
    }

    private fun selectVideoTrack(extractor: MediaExtractor): Int {
        for (index in 0 until extractor.trackCount) {
            val mimeType = extractor.getTrackFormat(index).getString(MediaFormat.KEY_MIME)
            if (mimeType?.startsWith("video/") == true) return index
        }
        return -1
    }

    private fun videoRotationDegrees(format: MediaFormat): Int {
        if (!format.containsKey(MediaFormat.KEY_ROTATION)) return 0
        return normalizeRotationDegrees(format.getInteger(MediaFormat.KEY_ROTATION))
    }

    private fun normalizeRotationDegrees(rotationDegrees: Int): Int {
        val normalized = ((rotationDegrees % 360) + 360) % 360
        return when (normalized) {
            in 45..134 -> 90
            in 135..224 -> 180
            in 225..314 -> 270
            else -> 0
        }
    }

    private fun drainVideo(
        extractor: MediaExtractor,
        decoder: MediaCodec,
        imageHandler: Handler,
    ) {
        val bufferInfo = MediaCodec.BufferInfo()
        var inputEnded = false
        var playbackStartMs = -1L

        while (!cancelled) {
            if (!inputEnded) {
                val inputIndex = decoder.dequeueInputBuffer(CODEC_TIMEOUT_US)
                if (inputIndex >= 0) {
                    val inputBuffer = decoder.getInputBuffer(inputIndex)
                    inputBuffer?.clear()
                    val sampleSize = inputBuffer?.let { extractor.readSampleData(it, 0) } ?: -1
                    if (sampleSize < 0) {
                        decoder.queueInputBuffer(
                            inputIndex,
                            0,
                            0,
                            0,
                            MediaCodec.BUFFER_FLAG_END_OF_STREAM,
                        )
                        inputEnded = true
                    } else {
                        decoder.queueInputBuffer(
                            inputIndex,
                            0,
                            sampleSize,
                            extractor.sampleTime,
                            extractor.sampleFlags,
                        )
                        extractor.advance()
                    }
                }
            }

            val outputIndex = decoder.dequeueOutputBuffer(bufferInfo, CODEC_TIMEOUT_US)
            if (outputIndex >= 0) {
                if (playbackStartMs < 0) {
                    playbackStartMs =
                        System.currentTimeMillis() - bufferInfo.presentationTimeUs / 1000L
                }
                waitUntilPresentationTime(playbackStartMs, bufferInfo.presentationTimeUs)
                val endOfStream =
                    bufferInfo.flags and MediaCodec.BUFFER_FLAG_END_OF_STREAM != 0
                decoder.releaseOutputBuffer(outputIndex, !endOfStream)
                if (endOfStream) {
                    waitForPendingVideoFrames(imageHandler)
                    decoder.flush()
                    extractor.seekTo(0, MediaExtractor.SEEK_TO_PREVIOUS_SYNC)
                    inputEnded = false
                    playbackStartMs = -1L
                }
            } else if (outputIndex == MediaCodec.INFO_OUTPUT_FORMAT_CHANGED) {
                Log.d(LOG_TAG, "Video decoder output format changed: ${decoder.outputFormat}")
            }
        }
    }

    private fun waitForPendingVideoFrames(imageHandler: Handler) {
        val imagesDrained = CountDownLatch(1)
        val framesSubmitted = CountDownLatch(1)
        imageHandler.post {
            imagesDrained.countDown()
            mainHandler.post { framesSubmitted.countDown() }
        }
        try {
            imagesDrained.await()
            framesSubmitted.await()
        } catch (error: InterruptedException) {
            Thread.currentThread().interrupt()
            cancelled = true
        }
    }

    private fun waitUntilPresentationTime(playbackStartMs: Long, presentationTimeUs: Long) {
        val delayMs = playbackStartMs + presentationTimeUs / 1000L - System.currentTimeMillis()
        if (delayMs <= 0) return
        try {
            Thread.sleep(delayMs)
        } catch (error: InterruptedException) {
            Thread.currentThread().interrupt()
            cancelled = true
        }
    }

    private fun submitHardwareFrame(
        width: Int,
        height: Int,
        dataSpace: Int,
        rotationDegrees: Int,
        hardwareBuffer: HardwareBuffer,
        releaseOwner: () -> Unit,
    ) {
        postHardwareFrameWhenSurfaceIsReady(
            HardwareFrame(
                width,
                height,
                dataSpace,
                rotationDegrees,
                hardwareBuffer,
                releaseOwner,
            ),
        )
    }

    private fun postHardwareFrameWhenSurfaceIsReady(frame: HardwareFrame) {
        var shouldPost = false
        synchronized(hardwareFrameLock) {
            pendingHardwareFrame?.close()
            pendingHardwareFrame = frame
            if (!hardwareFramePostScheduled) {
                hardwareFramePostScheduled = true
                shouldPost = true
            }
        }
        if (shouldPost) {
            mainHandler.post(::postPendingHardwareFrameOnMainThread)
        }
    }

    private fun postPendingHardwareFrameOnMainThread() {
        val frame = synchronized(hardwareFrameLock) {
            pendingHardwareFrame.also {
                pendingHardwareFrame = null
                if (it == null) hardwareFramePostScheduled = false
            }
        } ?: return

        if (cancelled) {
            frame.close()
            synchronized(hardwareFrameLock) {
                hardwareFramePostScheduled = false
            }
            return
        }
        if (!controller.isReady()) {
            mainHandler.postDelayed(
                ::postPendingHardwareFrameOnMainThread,
                SURFACE_RETRY_DELAY_MS,
            )
            synchronized(hardwareFrameLock) {
                if (pendingHardwareFrame != null) {
                    frame.close()
                } else {
                    pendingHardwareFrame = frame
                }
            }
            return
        }
        try {
            controller.postHardwareBuffer(
                frame.width,
                frame.height,
                frame.dataSpace,
                frame.rotationDegrees,
                frame.hardwareBuffer,
            )
        } finally {
            frame.close()
        }

        val shouldPostAgain = synchronized(hardwareFrameLock) {
            (pendingHardwareFrame != null).also { hardwareFramePostScheduled = it }
        }
        if (shouldPostAgain) {
            mainHandler.post(::postPendingHardwareFrameOnMainThread)
        }
    }

    private fun closePendingHardwareFrame() {
        synchronized(hardwareFrameLock) {
            pendingHardwareFrame?.close()
            pendingHardwareFrame = null
            hardwareFramePostScheduled = false
        }
    }

    private fun waitForAcquireFence(image: Image) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return

        try {
            image.fence.use { fence: SyncFence? ->
                if (fence != null && fence.isValid && !fence.awaitForever()) {
                    Log.w(LOG_TAG, "Media acquire fence reported an error")
                }
            }
        } catch (error: IOException) {
            Log.w(
                LOG_TAG,
                "Cannot get media acquire fence; continuing without explicit wait",
                error,
            )
        }
    }

    private class HardwareFrame(
        val width: Int,
        val height: Int,
        val dataSpace: Int,
        val rotationDegrees: Int,
        val hardwareBuffer: HardwareBuffer,
        val releaseOwner: () -> Unit,
    ) {
        fun close() {
            try {
                hardwareBuffer.close()
            } finally {
                releaseOwner()
            }
        }
    }

    private companion object {
        const val LOG_TAG = "MediaFrameSource"
        const val MAX_IMAGE_EDGE = 1920
        const val MAX_VIDEO_IMAGES = 6
        const val SURFACE_RETRY_DELAY_MS = 16L
        const val CODEC_TIMEOUT_US = 10_000L
    }
}
